File: Problems/Algorithms/695. Max Area of Island/max_area.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        n, m = len(grid), len(grid[0])
        visited = [[0 for _ in range(m)] for _ in range(n)]
        neighbors = [[0, 1], [0, -1], [1, 0], [-1, 0]]

        def dfs(grid: List[List[int]], visited: List[List[int]], r: int, c: int) -> int:
            if not self.is_valid(n, m, r, c) or visited[r][c] == 1 or grid[r][c] == 0:
                return 0

            visited[r][c] = 1
            ans = 1
            for neighbor in neighbors:
                ans += dfs(grid, visited, r+neighbor[0], c+neighbor[1])
            return ans

        best = 0
        for r in range(n):
            for c in range(m):
                best = max(best, dfs(grid, visited, r, c))
        return best

    @staticmethod
    def is_valid(n: int, m: int, r: int, c: int) -> bool:
        return r >= 0 and c >= 0 and r < n and c < m

File: Problems/Algorithms/695. Max Area of Island/MaxArea.java
class Solution {
    public int maxAreaOfIsland(int[][] grid) {
        int n = grid.length;
        int m = grid[0].length;
        int[][] visited = new int[n][m];
        
        int ans = 0;
        for (int r = 0; r < n; r++) {
            for (int c = 0; c < m; c++) {
                ans = Math.max(ans, dfs(grid, visited, n, m, r, c));
            }
        }
        
        return ans;
    }
    
    private int dfs(int[][] grid, int[][] visited, int n, int m, int r, int c) {
        if (!isValid(n, m, r, c) || grid[r][c] == 0 || visited[r][c] == 1) return 0;
        
        visited[r][c] = 1;
        return 1 + dfs(grid, visited, n, m, r+1, c) + dfs(grid, visited, n, m, r-1, c) + dfs(grid, visited, n, m, r, c+1) + dfs(grid, visited, n, m, r, c-1);
    }
    
    private boolean isValid(int n, int m, int r, int c) {
        return r >= 0 && c >= 0 && r < n && c < m;
    }
}

